blob/s3blob/s3blob.go

```diff
@@ -628,11 +663,33 @@ func unescapeKey(key string) string {
 // NewTypedWriter implements driver.NewTypedWriter.
 func (b *bucket) NewTypedWriter(ctx context.Context, key string, contentType string, opts *driver.WriterOptions) (driver.Writer, error) {
 	key = escapeKey(key)
-	uploader := s3manager.NewUploaderWithClient(b.client, func(u *s3manager.Uploader) {
-		if opts.BufferSize != 0 {
-			u.PartSize = int64(opts.BufferSize)
-		}
-	})
+
+	// Check if a cached uploader already exists.
+	b.RLock()
+	cachedUploader := b.cachedUploader
+	b.RUnlock()
+
+	var (
+		uploader    *s3manager.Uploader
+		uploaderKey = newUploaderKey(opts.BufferSize)
+	)
+	if cachedUploader != nil && uploaderKey == cachedUploader.key {
+		// If a cached uploader already exists *and* it has the same key/configuration
+		// as requested by the caller then reuse it.
+		uploader = cachedUploader.uploader
+	} else {
+		// Otherwise create a new uploader.
```

Comment:
There is a race here; two goroutines could enter this function, both get a `nil` `cachedUploader`, and then both create a new one. One of them will win to be saved. Using a helper function that grabs the lock for the entire scope of the function should fix this.